File: C++/ThreeSumClosest.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;



int threeSumClosest(vector<int> &num, int target) {
    int n = num.size();
    if (n < 3)
        return 0;
    int result = num[0] + num[1] + num[2];
    sort(num.begin(), num.end());
    for (int i = 0; i < n - 2; ++i)
    {
        for (;(i > 0) && (num[i] == num[i-1]); ++i);
        int start = i + 1;
        int end = n - 1;
      
        while (start < end) {
            int curSum = num[i] + num[start] + num[end];
            if (abs(curSum-target) < abs(result-target))
                result = curSum;
            if (curSum <= target) {
                if (curSum == target)
                    return curSum;
                for(start++; num[start] == num[start-1]; ++start);
            } else {
                for (end--; num[end] == num[end+1]; --end);
            }
        }
    }
    
    return result;
}

int main(int argc, char** argv)
{
    return 0;
}

File: Java/ThreeSumClosest.java
public class ThreeSumClosest {

    // O(n^2) time and O(1) space
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int closestSum = 0;
        int diff = Integer.MAX_VALUE;
        for (int i = 0; i < nums.length; ++i) {
            if ((i < nums.length-1) && (nums[i] == nums[i+1])) {
                continue;
            }
            int start = 0;
            int end = i - 1;
            int other = target - nums[i];
            while (start < end) {
                if ((start > 0) && (nums[start] == nums[start-1])) {
                    start++;
                    continue;
                }
                if ((end < i-1) && (nums[end] == nums[end+1])) {
                    end--;
                    continue;
                }
                int sum = nums[start] + nums[end];
                if (sum == other) {
                    return target;
                } else if (sum < other) {
                    start++;
                } else {
                    end--;
                }
                int curDiff = Math.abs(sum+nums[i]-target);
                if (curDiff < diff) {
                    diff = curDiff;
                    closestSum = sum+nums[i];
                }
            }
        }
        return closestSum;
    }
}